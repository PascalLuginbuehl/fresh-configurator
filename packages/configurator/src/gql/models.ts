export interface FlightControllerPartial {
  port: string;
  connection: { connecting: boolean; connected: boolean };
}

export interface ConnectionStatusMapping {
  parent: FlightControllerPartial;
}
