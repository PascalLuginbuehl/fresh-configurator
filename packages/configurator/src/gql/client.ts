import { ApolloClient, InMemoryCache } from "@apollo/client";
import {
  ports,
  isOpen,
  open,
  getAttitude,
  close,
  initialise,
  apiVersion,
  bytesRead,
  bytesWritten,
  packetErrors,
  getStatusExtended,
  getRcValues
} from "@fresh/msp";
import semver from "semver";
import {
  Resolvers,
  ConnectionStateQueryVariables,
  ConnectionStateDocument,
  ConnectionStateQuery,
  LogsQuery,
  LogsQueryVariables,
  LogsDocument
} from "./__generated__";
import config from "../config";
import { versionInfo } from "../util";

interface Context {
  client: ApolloClient<object>;
}

const cache = new InMemoryCache();

const selectedPort = cache.makeVar<string | null>(null);
// default baudrate
const selectedBaud = cache.makeVar<number>(115200);

const expertMode = cache.makeVar<boolean>(false);
const selectedTab = cache.makeVar<string | null>(null);

cache.policies.addTypePolicies({
  Configurator: {
    fields: {
      port: () => selectedPort(),
      baudRate: () => selectedBaud(),
      tab: () => selectedTab(),
      expertMode: () => expertMode()
    }
  }
});

const setConnectionState = (
  client: ApolloClient<object>,
  port: string,
  connecting: boolean,
  connected: boolean
): void =>
  client.writeQuery<ConnectionStateQuery, ConnectionStateQueryVariables>({
    query: ConnectionStateDocument,
    data: {
      __typename: "Query",
      device: {
        connection: {
          connected,
          connecting,
          __typename: "ConnectionStatus"
        },
        __typename: "FlightController"
      }
    },
    variables: {
      port
    }
  });

const log = (client: ApolloClient<object>, message: string): void => {
  const data = client.readQuery<LogsQuery, LogsQueryVariables>({
    query: LogsDocument
  });

  client.writeQuery<LogsQuery, LogsQueryVariables>({
    query: LogsDocument,
    data: {
      configurator: {
        logs: [
          ...(data?.configurator.logs ?? []).concat({
            time: new Date().toISOString(),
            message,
            __typename: "Log" as const
          })
        ],
        __typename: "Configurator"
      },
      __typename: "Query"
    }
  });
};

const resolvers: Resolvers = {
  Query: {
    ports: () => ports(),
    device: (_, { port }) => ({
      port,
      apiVersion: "",
      connection: {
        port,
        connecting: false,
        connected: false,
        __typename: "ConnectionStatus"
      },
      __typename: "FlightController"
    }),
    configurator: () => {
      const { os, version, chromeVersion } = versionInfo();
      return {
        port: selectedPort(),
        baudRate: selectedBaud(),
        tab: selectedTab(),
        expertMode: expertMode(),
        logs: [
          {
            time: new Date().toISOString(),
            message: `Running - OS: <strong>${os}</strong>, Chrome: <strong>${chromeVersion}</strong>, Configurator: <strong>${version}</strong>`,
            __typename: "Log"
          }
        ],
        __typename: "Configurator"
      };
    }
  },
  Mutation: {
    connect: async (_, { port, baudRate }, { client }: Context) => {
      if (isOpen(port)) {
        return true;
      }

      let closed = false;

      setConnectionState(client, port, true, false);

      await open(port, { baudRate }, () => {
        // on disconnect
        log(
          client,
          `Serial port <span class="message-positive">successfully</span> closed on ${port}`
        );
        setConnectionState(client, port, false, false);
        closed = true;
      });
      log(
        client,
        `Serial port <span class="message-positive">successfully</span> opened on ${port}`
      );

      try {
        await initialise(port);
        const version = apiVersion(port);
        log(client, `MultiWii API version: <strong>${version}</strong>`);

        if (semver.gte(version, config.apiVersionAccepted)) {
          setConnectionState(client, port, false, true);
          return true;
        }
        log(
          client,
          `MSP version not supported: <span class="message-negative">${version}</span>`
        );
      } catch (e) {
        console.log(e);
        log(
          client,
          `No configuration received within <span class="message-negative">5 seconds</span>, communication <span class="message-negative">failed</span>`
        );
      }

      // And only close the port if we are connecting
      if (!closed) {
        await close(port);
      }

      return false;
    },
    disconnect: async (_, { port }, { client }: Context) => {
      if (!isOpen(port)) {
        return true;
      }

      await close(port);
      setConnectionState(client, port, false, false);

      return true;
    },

    setTab: (_, { tabId }) => !!selectedTab(tabId),
    setConnectionSettings: (_, { port, baudRate }) => {
      selectedPort(port);
      if (typeof baudRate === "number") {
        selectedBaud(baudRate);
      }
      return true;
    },
    setExpertMode: (_, { enabled }) => !!expertMode(enabled),
    log: (_, { message }, { client }: Context) => {
      log(client, message);
      return true;
    }
  },

  FlightController: {
    attitude: ({ port }) =>
      getAttitude(port).then(values => ({ ...values, __typename: "Attitude" })),
    status: ({ port }) =>
      getStatusExtended(port).then(values => ({
        ...values,
        __typename: "Status"
      })),
    rc: ({ port }) => ({
      port,
      __typename: "RC"
    }),
    apiVersion: ({ port }) => apiVersion(port)
  },
  ConnectionStatus: {
    bytesRead: ({ port }) => bytesRead(port),
    bytesWritten: ({ port }) => bytesWritten(port),
    packetErrors: ({ port }) => packetErrors(port)
  },
  RC: {
    channels: ({ port }) => getRcValues(port)
  }
};

const client = new ApolloClient({
  cache,
  // generated resolvers are not compatible with apollo
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  resolvers: resolvers as any
});

export default client;
