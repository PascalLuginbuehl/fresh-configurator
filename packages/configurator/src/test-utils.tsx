import React from "react";
import { render as renderFunction, RenderResult } from "@testing-library/react";
import { ThemeProvider } from "./theme";

const AllTheProviders: React.FC = ({ children }) => (
  <ThemeProvider>{children}</ThemeProvider>
);

const customRender = (
  ui: Parameters<typeof renderFunction>[0],
  options?: Parameters<typeof renderFunction>[1]
): RenderResult => renderFunction(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
