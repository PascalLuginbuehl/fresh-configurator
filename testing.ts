import { ports, connect, getVoltages, close } from "./device";

(async () => {
    const availablePorts = await ports();
    console.log(availablePorts);

    const port = availablePorts[0].path;
    await connect(port);
    try {
        console.log(await getVoltages(port));
    } catch (e) {
        console.log(e);
        await close(port);
    }
})();