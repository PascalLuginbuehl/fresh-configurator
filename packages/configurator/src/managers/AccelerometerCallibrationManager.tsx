import React from "react";
import { useCallibrateAccelerometerMutation } from "../gql/__generated__";
import Button from "../components/Button";
import useConnectionState from "../hooks/useConnectionState";

const AccelerometerCallibrationManager: React.FC = () => {
  const { connection } = useConnectionState();
  const [calibrate, { loading }] = useCallibrateAccelerometerMutation({
    variables: {
      connection: connection ?? "",
    },
  });

  return (
    <Button
      data-testid="calibrate-acc-button"
      disabled={loading || !connection}
      onClick={() => calibrate()}
    >
      {loading ? "Calibrating..." : "Calibrate Accelerometer"}
    </Button>
  );
};

export default AccelerometerCallibrationManager;
