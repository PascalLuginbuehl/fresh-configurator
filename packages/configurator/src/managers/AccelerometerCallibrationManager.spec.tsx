import React from "react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";

import { render, fireEvent, waitFor } from "../test-utils";
import { CallibrateAccelerometerDocument } from "../gql/__generated__";
import AccelerometerCallibrationManager from "./AccelerometerCallibrationManager";

const mockConnectionId = "someconnectionid";

const clientResolvers = {
  Query: {
    configurator: () => ({
      connection: mockConnectionId,
      connecting: false,
    }),
  },
};

const mockCallibrateMutation = (
  connection: string,
  callback: () => void
): MockedResponse => ({
  request: {
    query: CallibrateAccelerometerDocument,
    variables: {
      connection,
    },
  },
  result: () => {
    callback();
    return {
      data: null,
    };
  },
  delay: 100,
});

describe("AccelerometerCallibrationManager", () => {
  it("should callibrate the device accelerometer when button is clicked", async () => {
    const callibrate = jest.fn();
    const { getByTestId } = render(
      <MockedProvider
        resolvers={clientResolvers}
        mocks={[mockCallibrateMutation(mockConnectionId, callibrate)]}
      >
        <AccelerometerCallibrationManager />
      </MockedProvider>
    );

    await waitFor(() =>
      expect(getByTestId("calibrate-acc-button")).toBeEnabled()
    );

    expect(getByTestId("calibrate-acc-button")).toMatchSnapshot();
    fireEvent.click(getByTestId("calibrate-acc-button"));

    await waitFor(() =>
      expect(getByTestId("calibrate-acc-button")).toBeDisabled()
    );
    expect(getByTestId("calibrate-acc-button")).toMatchSnapshot();

    await waitFor(() =>
      expect(getByTestId("calibrate-acc-button")).toBeEnabled()
    );
    expect(getByTestId("calibrate-acc-button")).toMatchSnapshot();
    expect(callibrate).toHaveBeenCalled();
  });
});
