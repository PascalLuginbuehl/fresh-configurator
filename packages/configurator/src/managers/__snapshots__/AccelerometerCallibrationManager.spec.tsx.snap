// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccelerometerCallibrationManager should callibrate the device accelerometer when button is clicked 1`] = `
.c0 {
  width: 100%;
  padding: 5px 0px 5px 0px;
  text-align: center;
  background-color: #ffbb00;
  border-radius: 4px;
  border: 1px solid #dba718;
  color: #000;
  font-weight: 600;
  font-size: 12px;
  line-height: 13px;
  display: block;
  -webkit-transition: all ease 0.2s;
  transition: all ease 0.2s;
}

.c0:hover:enabled {
  cursor: pointer;
  background-color: #ffcc3f;
  color: #000;
  text-shadow: 0px 1px rgba(255,255,255,0.25);
  -webkit-transition: all ease 0.2s;
  transition: all ease 0.2s;
}

.c0:disabled {
  background-color: #f1f1f1;
  border-color: silver;
  color: #ccc;
}

.c0:active {
  background-color: #ffcc3f;
  -webkit-transition: all ease 0s;
  transition: all ease 0s;
  box-shadow: inset 0px 1px 5px rgba(0,0,0,0.35);
}

<button
  class="c0"
  data-testid="calibrate-acc-button"
>
  Calibrate Accelerometer
</button>
`;

exports[`AccelerometerCallibrationManager should callibrate the device accelerometer when button is clicked 2`] = `
<button
  class="sc-AxjAm bnNGmx"
  data-testid="calibrate-acc-button"
  disabled=""
>
  Calibrating...
</button>
`;

exports[`AccelerometerCallibrationManager should callibrate the device accelerometer when button is clicked 3`] = `
<button
  class="sc-AxjAm bnNGmx"
  data-testid="calibrate-acc-button"
>
  Calibrate Accelerometer
</button>
`;
